src/validation.cpp

```diff
@@ -1290,6 +1319,130 @@ static void CheckForkWarningConditionsOnNewFork(CBlockIndex* pindexNewForkTip)
+ */
+void CChainState::MaybeAddNewHeaderCandidate(CBlockIndex* pindex, bool chain_ordered_insertion, bool no_descendants) {
+    if (!pindex->IsValid(BLOCK_VALID_TREE)) return; // We only want things that have a valid header tree
+
+    bool lower_work = chainActive.Tip() != nullptr && chainActive.Tip()->nChainWork > pindex->nChainWork;
+    if (lower_work) return; // We don't want things with less work than our current tip
+
+    bool parent_present = false;
+    if (pindex->pprev && setBlockIndexHeaderCandidates.count(pindex->pprev)) {
+        // If the parent is a previous candidate, then no parents of it could
+        // be candidates, either. This is the only thing we need to check by
+        // definition of chain_ordered_insertion, however even in the case of
+        // !chain_ordered_insertion, if this is true, no need to do a full
+        // parent scan (as no further-up parent can be a candidate, either).
+        setBlockIndexHeaderCandidates.erase(pindex->pprev);
+        parent_present = true;
+    } else if (!chain_ordered_insertion) {
+        // We are being called in a for(p : mapBlockIndex) loop, so can make no
+        // assumptions about existing entries. Scan all other entries to check
+        // if we're a descendant of some other candidate.
+        for (auto it = setBlockIndexHeaderCandidates.begin(); it != setBlockIndexHeaderCandidates.end() && (*it)->nChainWork < pindex->nChainWork; it++) {
+            if (pindex->GetAncestor((*it)->nHeight) == *it) {
+                // it should be removed - we only keep the tip of potential
+                // chains, not anything in them. At this point we should be
+                // consistent by adding pindex, there should be more more work
+                // to do here.
+                setBlockIndexHeaderCandidates.erase(it);
+                break;
+                parent_present = true;
+            }
+        }
+    }
+
+    if (!parent_present && !no_descendants) {
+        // Scan higher-work entries to check that we're not a parent of some
+        // other candidate(s). If a parent of ours was already present then we
+        // can be certain that no such child is also a candidate, so we can
+        // skip the whole scan.
+        for (auto it = setBlockIndexHeaderCandidates.rbegin(); it != setBlockIndexHeaderCandidates.rend() && (*it)->nChainWork > pindex->nChainWork; it++) {
+            if ((*it)->GetAncestor(pindex->nHeight) == pindex) {
+                // pindex is useless - even if there are other tips based on it
+                // which we want in setBlockIndexHeaderCandidates, we're not
+                // gonna find them here.
+                return;
+            }
+        }
+    }
+
+    setBlockIndexHeaderCandidates.insert(pindex);
+}
+
+const CBlockIndex* CChainState::GetBestHeader() {
+    LOCK(cs_main);
+    auto it = setBlockIndexHeaderCandidates.rbegin();
+    if (it == setBlockIndexHeaderCandidates.rend())
+        return nullptr;
```

Comment:
Numerous places call this function but no nullptr check is performed and dereferences are executed on the ptr.
Can this return null in any case where this function is called?

Reply:
That seems problematic indeed. There's an `assert` for this in https://github.com/bitcoin/bitcoin/blame/7b1ab22384d944f638c514f838e25b391204ef67/src/net_processing.cpp#L3308, but not in the calls from validation.cpp:
* in `NotifyHeaderTip()`: here it's compared with `pindexHeaderOld` which is a `nullptr` initially, but would non-null the next time. It's not dereferenced in that function, but it is passed on to `uiInterface.NotifyHeaderTip`. I could add an assert inside`if (fNotify)`
* in `IsCurrentForFeeEstimation()`: I could add an `assert` directly above the if statement that deferences it

The comment in `InvalidBlockFound` suggests that sometimes `setBlockIndexHeaderCandidates` can be empty, so the above wouldn't be enough.